firedrake/slate/slac/utils.py

```diff
@@ -136,6 +152,113 @@ def visit_Symbol(self, o, *args, **kwargs):
         return SymbolWithFuncallIndexing(o.symbol, o.rank, o.offset)
 
 
+def slate_to_gem(expression):
+    """ Method encapsulating stage 1 of tsslac. """
+    mapper, var2terminal = slate2gem(expression)
+    if not (type(mapper) == Indexed or type(mapper) == FlexiblyIndexed):
+        shape = mapper.shape if len(mapper.shape) != 0 else (1,)
+        mapper = Indexed(mapper, make_indices(len(shape)))
+    return list([mapper]), var2terminal
+
+
+@singledispatch
+def _slate2gem(expr, self):
+    raise AssertionError("Cannot handle terminal type: %s" % type(expr))
+
+
+@_slate2gem.register(sl.Tensor)
+@_slate2gem.register(sl.AssembledVector)
+def _slate2gem_tensor(expr, self):
+    shape = expr.shape if not len(expr.shape) == 0 else (1, )
+    name = f"T{len(self.var2terminal)}"
+    assert expr not in self.var2terminal.values()
+    var = Variable(name, shape)
+    self.var2terminal[var] = expr
+    return var
+
+
+@_slate2gem.register(sl.Block)
+def _slate2gem_block(expr, self):
+    child, = map(self, expr.children)
+
+    # get index of first block in a range of blocks
+    first_ind = ()
+    is_range = lambda index: (type(index) == range)
+    for index in expr._indices:
+        first_ind += (index.start,) if is_range(index) else (index,)
```

Comment:
I think this code would be simpler if we had a consistent index type in the `Block` constructor. Likely always turning either slices, individual indices, or something else into tuples (perhaps singleton tuples).

Reply:
Like this (fixing a bug in the block indexer)
```patch
diff --git a/firedrake/slate/slate.py b/firedrake/slate/slate.py
index a397cc7e..69e77324 100644
--- a/firedrake/slate/slate.py
+++ b/firedrake/slate/slate.py
@@ -65,27 +65,26 @@ class BlockIndexer(object):
         self.block_cache = {}
 
     def __getitem__(self, key):
-
-        key = list(as_tuple(key))
-
+        key = as_tuple(key)
         # Make indexing with too few indices legal.
-        key += [slice(None) for i in range(self.tensor.rank - len(key))]
-
+        key = key + tuple(slice(None) for i in range(self.tensor.rank - len(key)))
         if len(key) > self.tensor.rank:
             raise ValueError("Attempting to index a rank-%s tensor with %s indices."
                              % (self.tensor.rank, len(key)))
 
+        block_shape = tuple(len(V) for V in self.tensor.arg_function_spaces)
         # Convert slice indices to tuple of indices.
-        blocks = tuple(range(n)[k] if isinstance(k, slice) else k
-                       for k, n in zip(key, self.tensor.shape))
+        blocks = tuple(as_tuple(range(n)[k] if isinstance(k, slice) else k)
+                       for k, n in zip(key, block_shape))
 
+        if blocks == tuple(tuple(range(n)) for n in block_shape):
+            return self.tensor
         # Avoid repeated instantiation of an equivalent block
         try:
             block = self.block_cache[blocks]
         except KeyError:
             block = Block(tensor=self.tensor, indices=blocks)
             self.block_cache[blocks] = block
-
         return block
```

Then you can just write:

```python
def _slate2gem_block(expr, self):
    child, = map(self, expr.children)
    child_shapes = expr.children[0].shapes
    offsets = tuple(sum(shape[:idx]) for shape, (idx, *_)
                    in zip(child_shapes.values(), expr._indices))
    return view(child, *(slice(idx, idx+extent) for idx, extent in zip(offsets, expr.shape)))
```

Reply:
This is a lot neater, thanks.